libgnucash/engine/Account.cpp

```diff
@@ -4856,211 +4678,93 @@ xaccAccountGetLastNum (const Account *acc)
 void
 xaccAccountSetLastNum (Account *acc, const char *num)
 {
-    GValue v = G_VALUE_INIT;
-    g_return_if_fail(GNC_IS_ACCOUNT(acc));
-    g_value_init (&v, G_TYPE_STRING);
-
-    g_value_set_static_string (&v, num);
-    xaccAccountBeginEdit (acc);
-    qof_instance_set_path_kvp (QOF_INSTANCE (acc), &v, {"last-num"});
-    mark_account (acc);
-    xaccAccountCommitEdit (acc);
+    set_kvp_string_path (acc, { "last-num" }, num);
 }
 
 
 /********************************************************************\
 \********************************************************************/
 
-gboolean
-xaccAccountGetHigherBalanceLimit (const Account *acc,
-                                  gnc_numeric *balance)
+static bool
+get_balance_limit (const Account* acc, gnc_numeric* balance, const std::string& key,
+                   std::optional<std::optional<gnc_numeric>>& priv_limit)
 {
     g_return_val_if_fail (GNC_IS_ACCOUNT(acc), false);
+    g_return_val_if_fail (balance, false);
 
-    if (GET_PRIVATE(acc)->higher_balance_limit.has_value())
+    if (priv_limit.has_value())
     {
-        *balance = GET_PRIVATE(acc)->higher_balance_limit.value();
-
-        if (gnc_numeric_check (*balance) == 0)
-            return true;
-        else
-            return false;
+        auto limit = *priv_limit;
+        if (limit) *balance = *limit;
+        return limit.has_value();
     }
-    else
+    else if (auto priv_limit = get_kvp_gnc_numeric_path (acc, { KEY_BALANCE_LIMIT, key }))
```

Comment:
Didn't notice this before: You're masking the paramerter, you should change the name of this `priv_limit`.